### DISCUSSION TITLE
Best way of adding a Shell to an Paketo-Spring-Boot-Image

### DISCUSSION THREAD

**Discussion by yannick1608** [paketo-buildpacks/feedback#314] [Q&A] [answered]
Hello everyone,

We're currently building out Spring Boot images with the Spring Boot Maven Plugin using the `paketobuildpacks/run-jammy-tiny` image, which comes without any pre-installed shell, and import Buildpacks manually. Here's the list of out manually imported builders:

- `paketobuildpacks/ca-certificates`
- `paketobuildpacks/bellsoft-liberica`
- `paketobuildpacks/syft`
- `paketobuildpacks/executable-jar`
- `paketobuildpacks/spring-boot`

While this setup works great for our production environment, we sometimes find it challenging during development, especially for debugging purposes, where having access to a shell within the container would be beneficial.

I see a few potential solutions:

1. Extending the `paketobuildpacks/run-jammy-tiny` image to include a shell, though this might not be the most elegant solution.
2. Using a different run-image that already includes a shell. Does anyone have recommendations for such an image that would be suitable for development?
3. Alternatively, is there a specific buildpack that could install a shell within the container?

Is there another way to add a shell? Or, what would be the cleanest way to add a shell? Additionally, is there an example that demonstrates how to install the shell in such a way that it can be started alongside the main process (the Spring Boot application) when needed?

We are really grateful for the headless image provided by Paketo, as it's perfect for production. However, for development, having a shell as an extra option would streamline our workflow.

Looking forward to hearing your thoughts and any other suggestions you might have!

**Comment by dmikusa** [comment, marked as answer]
I can totally sympathize here. Not having the shell is an enormous pain.

> Using a different run-image that already includes a shell. Does anyone have recommendations for such an image that would be suitable for development?

If you need one, the simplest thing is to rebuild and switch from the `paketobuildpacks/builder-jammy-tiny` builder to the `paketobuildpacks/builder-jammy-base` builder (similar switch to the run image, `paketobuildpacks/run-jammy-base`). The base image has a slightly larger set of packages (about 50M larger). It includes bash, so you have a shell accessible.

> Alternatively, is there a specific buildpack that could install a shell within the container?

This would be interesting, but I don't know of a buildpack that does this. I think the challenge would be finding a shell that is statically compiled so that you don't need to also install a bunch of other stuff too.

The tiny image is also tricky because there's no package management tools either, so you can't just `apt install` (or use the apt-buildpack) either.

Hope that helps!
